File: dynamic_programming/staircase.py
from itertools import combinations


def solution(n):
    l = [] 

    #List l contains all the possible steps that can be used
    for i in range(1,n):
        l.append(i)

    #sol continas all the possible combinations of intergers of list l
    sol = []
    
    #How many possible ways can you combine steps to produce a sum
    for k in range(2,len(l)+1):
        #return all possible subsets with length k
        for m in combinations(l,k):
            sol.append(m)
    
    print("sol ", sol)

    final = []
    
    #Iterate through sol, and only append combinations with sum == height
    for z in (sol):
        if sum(z) == n :
            final.append(z)
    
    print("Final ", final)

    steps = len(final)
    
    return (steps)


n= 5
print("no of possible combinations: ", solution(n))

File: dynamic_programming/staircase.cpp
#include <bits/stdc++.h>
#include <algorithm>

//You have to build a staircase in such a way that, each type of staircase should consist of 
// 2 or more steps. No two steps are allowed to be at the same height — each step must be lower 
// than the previous one. All steps must contain at least one brick. A step’s height is 
// classified as the total amount of bricks that make up that step.

//returns number of ways stairs can be put together but 1-0
int staircaseFunc(int N){
    //the knapsack contains steps from 1 up to N-1
    
    std::vector<int> num_combi(N+1, 0);

    num_combi[0] = 1;

    //Iterate through the number of bricks
    for (int i = 1; i <= N; i++){
        //Iterate through the max height of the bricks
        for (int j = N; j >= i; j--){
            num_combi[j] += num_combi[j-i];
        }
    }

    return num_combi[N] - 1;

}


int main() {
    int N = 200;

    int num_ways = staircaseFunc(N);

    std::cout << "Number of ways: " << num_ways << std::endl;

    return 0;
}
